marshal/marshal.go

```diff
@@ -315,62 +315,72 @@ func (m *Marshaler) ClaimPartition(topicName string, partID int) bool {
 // Heartbeat will send an update for other people to know that we're still alive and
 // still owning this partition. Returns an error if anything has gone wrong (at which
 // point we can no longer assert we have the lock).
-func (m *Marshaler) Heartbeat(topicName string, partID int, CurrentOffset int64) error {
+func (m *Marshaler) Heartbeat(topicName string, partID int, offset int64) error {
 	topic, err := m.getClaimedPartitionState(topicName, partID)
 	if err != nil {
 		return err
 	}
 
+	// Attempt to commit offset, this is best-effort and we don't care if it fails
+	// since the canonical storage is in the heartbeat
+	if err := m.CommitOffsets(topicName, partID, offset); err != nil {
+		log.Warningf("[%s:%d] failed to commit offset during heartbeat: %s",
+			topicName, partID, err)
+	}
+
 	// All good, let's heartbeat
 	cl := &msgHeartbeat{
 		msgBase:       *m.msgBase(topicName, partID),
-		CurrentOffset: CurrentOffset,
+		CurrentOffset: offset,
 	}
 	_, err = m.cluster.producer.Produce(MarshalTopic, int32(topic.claimPartition),
 		&proto.Message{Value: []byte(cl.Encode())})
 	if err != nil {
+		log.Errorf("[%s:%d] failed to send heartbeat message to Kafka: %s",
+			topicName, partID, err)
 		return fmt.Errorf("Failed to produce heartbeat to Kafka: %s", err)
 	}
 
-	err = m.CommitOffsets(topicName, partID, CurrentOffset)
-	return err
+	return nil
 }
 
 // ReleasePartition will send an update for other people to know that we're done with
 // a partition. Returns an error if anything has gone wrong (at which
 // point we can no longer assert we have the lock).
-func (m *Marshaler) ReleasePartition(topicName string, partID int, CurrentOffset int64) error {
+func (m *Marshaler) ReleasePartition(topicName string, partID int, offset int64) error {
 	topic, err := m.getClaimedPartitionState(topicName, partID)
 	if err != nil {
 		return err
 	}
 
+	// Commit our offset first; if this fails, we can still try to release,
+	// but we should advise
```

Comment:
Suppose that `Heartbeat` gets called as a result of a "rewind" operation, and `CommitOffsets` fails but producing the heartbeat message succeeds.  Would subsequently shutting down the consumer process for an hour effectively undo the effect of the rewind operation?

This shouldn't be a concern for Telescope because we have a 5 minute limit on how long a process can be down before it loses its coord claim, but could future clients potentially be surprised by this behavior?